pahfit/fitter.py

```diff
@@ -0,0 +1,209 @@
+        pass
+
+    @abstractmethod
+    def finalize_model(self):
+        """Process the registered features and prepare for fitting.
+
+        The register functions below allow adding individual features.
+        The exact implementation of how features are added, and
+        finalized in to a single fittable model, depend on the
+        underlying implementation.
+
+        """
+        pass
+
+    @abstractmethod
+    def register_starlight(self, name, temperature, tau):
+        """Register a starlight feature.
+
+        The exact representation depends on the implementation, but the
+        meaning of the parameters should be equivalent.
+
+        Parameters
+        ----------
+        name : str
+            Unique name. Will be used to allow retrieval of the results
+            after the fitting.
+
+        temperature : array of size 3
+            Blackbody temperature, given as [value, lower_bound,
+            upper_bound]. The bounds assume the same system as the
+            features table: if they are masked, the parameter will be
+            fixed, while +inf or -inf mean unbounded.
+
+        tau : array of size 3
+            Analogously, used as power.
+
+        """
+        pass
+
+    @abstractmethod
+    def register_dust_continuum(self, name, temperature, tau):
+        """Register a dust continuum feature."""
+        pass
+
+    @abstractmethod
+    def register_line(self, name, power, wavelength, fwhm):
+        """Register an emission line feature.
+
+        Typically a Gaussian profile.
+
+        """
+        pass
+
+    @abstractmethod
+    def register_dust_feature(self, name, power, wavelength, fwhm):
+        """Register a dust feature.
+
+        Typically a Drude profile.
+
+        """
+        pass
+
+    @abstractmethod
+    def register_attenuation(self, name, tau):
+        """Register the S07 attenuation component.
+
+        Other types of attenuation might be possible in the future. Is
+        multiplicative.
+
+        """
+        pass
+
+    @abstractmethod
+    def register_absorption(self, name, tau, wavelength, fwhm):
+        """Register an absorption feature.
+
+        Typically a Drude profile. Is multiplicative.
+
+        """
+        pass
+
+    @abstractmethod
+    def evaluate_model(self, xz):
+        """Evaluate the model at the given wavelengths.
+
+        Parameters
+        ----------
+        xz : array
+            Rest frame wavelengths in micron
+
+        Returns
+        -------
+        yz : array
+            Rest frame flux in internal units
+
+        """
+        pass
+
+    @abstractmethod
+    def fit(self, xz, yz, uncz, maxiter=1000):
+        """Perform the fit using the framework of the subclass.
+
+        Fitter is unit agnostic, and deals with the numbers the Model
+        tells it to deal with. In practice, the input spectrum is
+        expected to be in internal units, and corrected for redshift
+        (models operate in the rest frame).
```

Comment:
`fit` occurs in the rest frame.  If we fit for wavelength, are we prepared to "undo" the redshift correction?

Reply:
For me, fitting the wavelength makes the most sense in two cases
1. a small correction to the rest frame wavelength of a line (to deal with e.g. small wavelength calibration issues)
2. determining the rest frame wavelength of a dust feature (which can shift for some of them, depending on the physics)

These cases have not much to do with redshift (other than the fact that we need to shift the observed data to the rest frame, to make such fitting work). When there are actual velocities causing the lines to shift, things are a bit more abiguous. 

But no, at the moment there is no utility to convert the rest frame values back to the observer frame. It might be useful to implement this somewhere, as a utility for visualizations (direct comparison to the data).

Reply:
Sorry I didn't word that well.  I meant "correcting" for redshift is something we ourselves do in `Model`, yes?  I.e. the user doesn't have to `lam_rest = lam/(1+z)` which this doc makes it sound like.  Do we accept a `z` on a model and/or in a `Spectrum1D`?

Reply:
Sorry I was confused about that.